nemo/collections/tts/modules/transformer_dec24.py

```diff
@@ -0,0 +1,691 @@
+            attn_prob = attn_prob.masked_fill(mask, 0.0)
+
+        attn_prob = self.dropout(attn_prob)
+
+        y = torch.matmul(attn_prob, v)
+        y = y.transpose(1, 2).contiguous().view(B, T, -1)
+
+        return y, [attn_prob, _attn_score]
+
+    def forward(
+        self, query, query_mask=None, memory=None, memory_mask=None, attn_prior=None, dump_attention=False, idx=None
+    ):
+        """
+        all inputs should be (B, T, C)
+        query_mask (T1, T1)
+        memory_mask (B, T2)
+        attn_prior (T1, T2)
+        """
+
+        if self.use_flash_attention:
+            attn_prob = []
+            y = self.attn_flash(query, query_mask, memory, memory_mask, idx)
+        else:
+            y, attn_prob = self.attn_naive(query, query_mask, memory, memory_mask, attn_prior, dump_attention, idx=idx)
+
+        y = self.dropout(self.o_net(y))
+
+        return y, attn_prob
+
+
+class TransformerBlock(nn.Module):
+    def __init__(
+        self,
+        d_model,
+        n_heads,
+        kernel_size,
+        p_dropout,
+        context_xattn,
+        has_xattn,
+        remove_self_attention=False,
+        is_causal=True,
+        apply_norm_to_cond=True,
+        layer_norm_method='pre',
+        use_layer_scale=False,
+        layer_scale_init=1e-1,
+        use_flash_self_attention=True,
+        use_flash_x_attention=True,
+        deterministic=False,
+        pos_emb={"name": "alibi"},
+        max_length_causal_mask=4096,
+    ):
+        super(TransformerBlock, self).__init__()
+        """
+        T5-ish
+        """
+        self.layer_norm_method = layer_norm_method
+        self.has_xattn = has_xattn
+        self.remove_self_attention = remove_self_attention
+        self.use_layer_scale = use_layer_scale
+
+        if not self.remove_self_attention:
+            self.norm_self = LayerNorm(d_model, use_bias=False)
+            self.self_attention = Attention(
+                n_heads=n_heads,
+                d_model=d_model,
+                p_dropout=p_dropout,
+                is_self_attention=True,
+                use_flash_attention=use_flash_self_attention,
+                deterministic=deterministic,
+                pos_emb=pos_emb,
+                max_length_causal_mask=max_length_causal_mask,
+            )
+            if self.use_layer_scale:
+                self.layer_scale_self_attn = LayerScale(d_model, init=layer_scale_init)
+            else:
+                self.layer_scale_self_attn = nn.Identity()
+
+        if self.has_xattn:
+            self.apply_norm_to_cond = apply_norm_to_cond
+            self.norm_xattn_query = LayerNorm(d_model, use_bias=False)
+            params = context_xattn['params']
+            cross_attention = Attention(
+                n_heads=params['n_heads'],
+                d_model=d_model,
+                p_dropout=p_dropout,
+                is_causal=False,
+                is_self_attention=False,
+                d_memory=params['d_heads'],
```

Comment:
Should rename `d_heads` in params to `d_memory` here. `d_memory` is supposed to be the dimension of the context information for cross attention. `d_heads` refers to the size of each attention head, but which this code hardcodes to be `d_memory // n_heads`.